overlord/servicestate/quota_control.go

```diff
@@ -44,6 +52,42 @@ func init() {
 	checkSystemdVersion()
 }
 
+func checkMemoryCGroupEnabled() {
+	memoryCGroupError = memoryCGroupEnabled()
+}
+
+var CGroupsFilePath = "/proc/cgroups"
+
+//	since the control groups can be enabled/disabled without the kernel config the only
+//	way to identify the status of memory control groups is via /proc/cgroups
+//	"cat /proc/cgroups | grep memory" returns the active status of memory control group
+//	and the 3rd parameter is the status
+//	0 => false => disabled 1 => true => enabled
+func memoryCGroupEnabled() error {
+	cgroupsFile, errFileOpen := os.Open(CGroupsFilePath)
+	if errFileOpen != nil {
+		return fmt.Errorf("internal error: cannot open cgroups file")
+	}
+	defer cgroupsFile.Close()
+	scanner := bufio.NewScanner(cgroupsFile)
+	scanner.Split(bufio.ScanLines)
+	for scanner.Scan() {
+		line := scanner.Text()
+		if strings.HasPrefix(line, "memory\t") {
+			isMemoryEnabled, errParseValue := strconv.ParseBool(strings.Fields(line)[3])
+			if errParseValue != nil {
+				return fmt.Errorf("cannot parse memory control group status")
+			}
+			if !isMemoryEnabled {
+				return fmt.Errorf("memory cgroup disabled on this system")
+			} else {
+				return nil
+			}
+		}
+	}
+	return fmt.Errorf("internal error: cannot parse memory control group status")
```

Comment:
Hm I'm not sure whether if the controller is actually disabled kernel config it would even appear on the list, in which case we'd not hit the prefix match case and incorrectly indicate that /proc/cgroup content cannot be parsed.

Reply:
If disabled it appears on the list as such but it's possible to leave it completely out of the config in which case the memory cgroup is not populated in /proc/cgroups. Updated the error message to reflect that as it's the only way to reach the end without obstruction.